hibernate-core/src/main/java/org/hibernate/engine/jdbc/env/internal/JdbcEnvironmentInitiator.java

```diff
@@ -73,105 +71,84 @@ public JdbcEnvironment initiateService(Map<String, Object> configurationValues,
+				() -> {
+					final Integer value = ConfigurationHelper.getInteger(
+							AvailableSettings.DIALECT_DB_MINOR_VERSION,
+							configurationValues
+					);
+					if ( value != null ) {
+						DeprecationLogger.DEPRECATION_LOGGER.deprecatedSetting(
 								AvailableSettings.DIALECT_DB_MINOR_VERSION,
-								configurationValues
+								AvailableSettings.JAKARTA_HBM2DDL_DB_MINOR_VERSION
 						);
-						if ( value != null ) {
-							DeprecationLogger.DEPRECATION_LOGGER.deprecatedSetting(
-									AvailableSettings.DIALECT_DB_MINOR_VERSION,
-									AvailableSettings.JAKARTA_HBM2DDL_DB_MINOR_VERSION
-							);
-						}
-						return value;
-					},
-					() -> 0
-			);
-			return new JdbcEnvironmentImpl( registry, dialectFactory.buildDialect(
-					configurationValues,
-					() -> new DialectResolutionInfo() {
-						@Override
-						public String getDatabaseName() {
-							return (String) dbName;
-						}
-
-						@Override
-						public String getDatabaseVersion() {
-							return dbVersion;
-						}
-
-						@Override
-						public int getDatabaseMajorVersion() {
-							return dbMajorVersion;
-						}
-
-						@Override
-						public int getDatabaseMinorVersion() {
-							return dbMinorVersion;
-						}
-
-						@Override
-						public String getDriverName() {
-							return "";
-						}
-
-						@Override
-						public int getDriverMajorVersion() {
-							return 0;
-						}
-
-						@Override
-						public int getDriverMinorVersion() {
-							return 0;
-						}
-
-						@Override
-						public String getSQLKeywords() {
-							return "";
-						}
 					}
-			) );
+					return value;
+				}
+		);
+
+		String explicitDatabaseVersion = NullnessHelper.coalesceSuppliedValues(
+				() -> (String) configurationValues.get( AvailableSettings.JAKARTA_HBM2DDL_DB_VERSION ),
+				() -> {
+					final Object value = configurationValues.get( AvailableSettings.DIALECT_DB_VERSION );
+					if ( value != null ) {
+						DeprecationLogger.DEPRECATION_LOGGER.deprecatedSetting(
+								AvailableSettings.DIALECT_DB_VERSION,
+								AvailableSettings.JAKARTA_HBM2DDL_DB_VERSION
+						);
+					}
+					return (String) value;
+				}
+				,
+				() -> {
+					if ( _explicitDatabaseMajorVersion != null ) {
+						return _explicitDatabaseMinorVersion == null ? _explicitDatabaseMajorVersion.toString() : (_explicitDatabaseMajorVersion + "." + _explicitDatabaseMinorVersion);
+					}
+					return null;
+				}
+		);
+
+		Integer explicitDatabaseMajorVersion;
+		Integer explicitDatabaseMinorVersion;
+		if ( _explicitDatabaseMajorVersion == null && explicitDatabaseVersion != null ) {
+			final String[] parts = explicitDatabaseVersion.split( "\\." );
+			final int potentialMajor = Integer.parseInt( parts[0] );
+			if ( parts.length > 1 ) {
+				explicitDatabaseMinorVersion = _explicitDatabaseMinorVersion == null ? Integer.parseInt( parts[1] ) : _explicitDatabaseMinorVersion;
+			}
```

Comment:
You forgot the most important parts, which is the exception handling. What was wrong with the code I posted before? We never use this kind of variable naming with prefixing by `_`, so I'd rather not do this here. It's fine to have these variables (`explicitDatabaseMajorVersion` and `explicitDatabaseMinorVersion`) be mutable here.

Reply:
i had to adapt it because the lambda's complain about variables not being final, and not being properly initialized and such ... and one of the causes of that was the try/catch. My change may not have been the best way to resolve these issues though, if you want to have a go it putting your original code block in there and resolving those issues more adequately i would certainly appreciate and learn from it.
For the record, i didn't change your code just for the sake of changing it